File: src/cmcandy/java_language_answers/_0053.java
package cmcandy.java_language_answers;

/**
 * Created by CHEN on 2020/5/10.
 */
public class _0053 {
    static class Solution {
        public int maxSubArray(int[] nums) {
            //贪心算法和动态规划
            //这里是动态规划

            //获取数组长度
            int len = nums.length;
            int[] tmp = new int[len];
            int max = nums[0];

            tmp[0] = nums[0];
            for(int i = 1;i<=len-1;i++){
                if(tmp[i-1]>0) tmp[i] = nums[i]+tmp[i-1];
                else tmp[i] =nums[i];
                if (max<tmp[i])max=tmp[i];
            }
            return max;
        }
    }
}

File: src/cmcandy/C_Language_Answers/_0053.cpp
#include <iostream>
#include <vector>
#include <limits.h>
using namespace ::std;

class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int max = INT_MIN, tmp = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            tmp = tmp <= 0 ? nums[i] : nums[i] + tmp;
            max = max > tmp ? max : tmp;
        }
        return max;
    }
};